src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,359 @@
+
+import android.content.Context;
+
+/**
+ * Internal class handling the detailed acquiretoken silent logic, including cache lookup and also
+ * interact with web request.
+ */
+class AcquireTokenSilentHandler {
+    private static final String TAG = AcquireTokenSilentHandler.class.getSimpleName();
+    
+    private final Context mContext;
+    private final TokenCacheAccessor mTokenCacheAccessor;
+    private final AuthenticationRequest mAuthRequest;
+    
+    private boolean mAttemptedWithFRT = false;
+    private TokenCacheItem mMrrtTokenCacheItem;
+
+    /**
+     * TODO: need to remove. {@link HttpUrlConnectionFactory} provides the possibility to 
+     * mock the real connection. Needs to update the class to make different response based on 
+     * post message. 
+     */
+    private IWebRequestHandler mWebRequestHandler = null;
+    
+    /**
+     * Constructor for {@link AcquireTokenSilentHandler}. 
+     * {@link TokenCacheAccessor} could be null. If null, won't handle with cache. 
+     */
+    AcquireTokenSilentHandler(final Context context, final AuthenticationRequest authRequest,
+            final TokenCacheAccessor tokenCacheAccessor) {
+        if (context == null) {
+            throw new IllegalArgumentException("context");
+        }
+        
+        if (authRequest == null) {
+            throw new IllegalArgumentException("authRequest");
+        }
+        
+        mContext = context;
+        mAuthRequest = authRequest;
+        mTokenCacheAccessor = tokenCacheAccessor;
+
+        mWebRequestHandler = new WebRequestHandler();
+    }
+    
+    /**
+     * Request for access token by looking up cache. . 
+     */
+    AuthenticationResult getAccessToken() throws AuthenticationException {
+        // If mTokenCacheAccessor is null, won't handle with token cache lookup. 
+        if (mTokenCacheAccessor == null) {
+            return null;
+        }
+        
+        TokenCacheItem tokenCacheItem = mTokenCacheAccessor.getRegularTokenCacheItemWithAT(mAuthRequest.getResource(), 
+                mAuthRequest.getClientId(), getUser());
+        
+        if (tokenCacheItem == null 
+                && lookUpCachedTokenWithNoUserForADFS()) {
+            // ADFS doesn't return idtoken back. If app is talking to ADFS server, and provide us the userId
+            // or displayableId, since we didn't get idtoken back, we don't store user as part of key in this
+            // case. Try again to find token with no user.
+            tokenCacheItem = mTokenCacheAccessor.getRegularTokenCacheItemWithAT(mAuthRequest.getResource(), 
+                    mAuthRequest.getClientId(), null);
+            // If we still don't have anything to use then we should try to find if we have a MRRT that matches
+            if (tokenCacheItem == null) {
+                return tryMRRT();
+            }
+        }
+        
+        if (isValidCache(tokenCacheItem)) {
+            if (isUserMisMatch(tokenCacheItem)) {
```

Comment:
Done. 

---
In reply to: [62919328](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r62919328) [](ancestors = 62919328)